akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/Effect.scala

```diff
@@ -45,11 +45,7 @@ object Effect {
   final case class AskInitiated[Req, Res, T](
       target: RecipientRef[Req],
       responseTimeout: FiniteDuration,
-      responseClass: Class[Res])(
-      replyToRef: ActorRef[Res],
-      val askMessage: Req,
-      asker: ActorRef[T],
-      mapResponse: Try[Res] => T)
+      responseClass: Class[Res])(val askMessage: Req, asker: ActorRef[T], mapResponse: Try[Res] => T)
```

Comment:
Why wouldn't the askMessage be part of equality for the effect? That seems natural for me, two initiated asks sending different requests are not very likely the same effect.

I also think there is something to improve with the design around the response here. The main reason that made us pull ask into the actor context is that the response mapping is run on the calling thread/inside the actor and any error there is handled as if it was an actor message, responding here requires the test to call the mapper and then send the result. 

I would prefer that responding is done through one of two methods on the effect `def respond(res: Res): Unit` and `def timeout(): Unit` that will make the behavior run the mapper and receive the result so that there result can be inspected. It could probably be implemented through the logic in the testkit already there for adapters/adapting messages.

It's still good to be able to run the response mapper directly to verify how it behaves, but I think that should  be a bonus feature rather than the way to drive the ask response.

Reply:
The rationale for excluding `askMessage` is that it will very likely contain the `replyTo` as a field which is considered for equality (virtually certain for Scala case class message, somewhat less certain for POJO messages given that the Java code I've seen tends to just use reference equality (someone who cared enough to implement `equals` on their POJO messages would probably consider `replyTo` for equality) and that ref shouldn't necessarily be something predictable (though strict parametric reasoning would lead one to the `ignoreRef` being the only possible `RecipientRef[Any]`).

Since `Effect` equality really only comes into play for comparing against an `Effect` constructed via the DSLs, that would require duplicating the `createRequest` function in the test for passing it to the DSL builder (since I don't see a way to get the `ignoreRef` in the test case).  Without the `mapResponse` function, of course, the DSL built `Effect` is useless for testing anything beyond the ask message was sent.

So it seems to come down to a choice between these two styles:

```scala
val askee = TestInbox[AskeeCommand]()

testKit.run(someCommand)

val AskeeCommand(something, somethingElse, _) = askee.receiveMessage()  // ignores replyTo
something shouldBe mumbo
somethingElse shouldBe jumbo

val askEffect = testKit.expectEffectType[AskInitiated[AskeeCommand, AskeeReply, Command]]()
askEffect.target shouldBe askee.ref      // arguably tested by the receiveMessage
askEffect.responseTimeout shouldBe 10.seconds
// can respond/timeout, test the actual adaptation function etc.
```

vs.

```scala
val askee = TestInbox[AskeeCommand]()

testKit.run(someCommand)

testKit.expectEffect(
  // hypothetical API considering askMessage for equality
  Effects.askInitiated[AskeeCommand, AskeeReply, Command](
    askee,
    10.seconds,
    classOf[AskeeReply],
    AskeeCommand(mumbo, jumbo, _)
  )
)
// equivalent to the test above, modulo not being able to do anything which needs the real effect from the testkit
```

or

```scala
val askee = TestInbox[AskeeCommand]()

testKit.run(someCommand)

val askEffect = testKit.expectEffectType[AskInitiated[AskeeCommand, AskeeReply, Command]]()
askEffect shouldBe
  Effects.askInitiated[AskeeCommand, AskeeReply, Command](
    askee,
    10.seconds,
    classOf[AskeeReply],
    AskeeCommand(mumbo, jumbo, _)
  )
// fully equivalent to the first example, have the real effect
```

Reply:
Good point, I did not think of the `replyTo` field. Let's keep equality like this then.

Reply:
Since the `Effects` DSL won't have access to the `ignoreRef` (which depends on the name of the `ActorSystemStub`) to inject, I guess we could have a static dummy `replyTo` ref for asks.

As for using the message adaptation logic, that would wrap the adapted response into an `AdaptMessage` which is `private[akka]` (and not a `T`), so that would break `selfInbox` assertions... since the adapter is available, I guess that's not necessarily a loss (and I would probably tend to not even bother with testing `respond`/`timeout`, so I guess it's no skin off my back).  It would force a `respond`/`timeout` to be followed with a `runOne`, which we could get around by passing the `BehaviorTestKitImpl` to the `EffectfulActorContext` (since I think the BTK is the only thing that uses the `EffectfulActorContext`?) and having `respond`/`timeout` `runOne`... though that could get hairy when dealing with, e.g., a message which does an ask and a self-send.

OTOH, if the BTKI is threaded through to the `EffectfulActorContext`, that would allow `respond`/`timeout` to `signal(PostStop)` on an exception and throw the exception to the test?

Reply:
...alternatively, if the test kit is threaded through to the `EffectfulActorContext`, `respond`/`timeout` could be a `run(AdaptedMessage(...))` which would synchronously propagate the exception.  In the ask and self-send situation, one would have to manually `runOne` and then `respond` (this might be enough discouragement around doing that sort of thing: depending on the self-send being in the mailbox before the ask response seems like something to discourage).